### PR TITLE
OperatorDef <==> NodeProto Conversion

### PR DESCRIPTION
**Pull request by houseroad**
Operator level proto conversion between (new) torch proto and (old) caffe2 proto.

### CODE REVIEW

File: caffe2/proto/torch_pb.h [hunk 1]
```diff
@@ -0,0 +1,3 @@
+#pragma once
```
Comment (BIT-silence) [inline caffe2/proto/torch_pb.h]: https://google.github.io/styleguide/cppguide.html#The__define_Guard

#ifndef CAFFE2_PROTO_TORCH_PB_H_
#define CAFFE2_PROTO_TORCH_PB_H_

...

#endif // CAFFE2_PROTO_TORCH_PB_H_

File: caffe2/python/pybind_state.cc [hunk 2]
```diff
@@ -1527,6 +1534,46 @@ void addGlobalMethods(py::module& m) {
...
+  });
+  m.def("operator_def_to_node_proto", [](py::bytes op_str) -> py::bytes {
+    OperatorDef op_def;
+    CAFFE_ENFORCE(
+      ParseProtoFromLargeString(op_str.cast<std::string>(), &op_def));
+    ::torch::NodeProto node;
+    OperatorDefToNodeProto(op_def, &node);
+    std::string ret;
+    node.SerializeToString(&ret);
+    return ret;
```
Comment (BIT-silence) [inline caffe2/python/pybind_state.cc]: return node.SerializeAsString();

https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.message_lite#MessageLite.SerializeAsString.details

File: caffe2/python/pybind_state.cc [hunk 3]
```diff
@@ -1527,6 +1534,46 @@ void addGlobalMethods(py::module& m) {
...
+  });
+  m.def("node_proto_to_operator_def", [](py::bytes node_str) -> py::bytes {
+    ::torch::NodeProto node_proto;
+    CAFFE_ENFORCE(
+      ParseProtoFromLargeString(node_str.cast<std::string>(), &node_proto));
+    OperatorDef op_def;
+    NodeProtoToOperatorDef(node_proto, &op_def);
+    std::string ret;
+    op_def.SerializeToString(&ret);
+    return ret;
```
Comment (BIT-silence) [inline caffe2/python/pybind_state.cc]: ditto.

File: caffe2/utils/proto_convert.cc [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    attr->set_type(::torch::AttributeProto_AttributeType::
+                       AttributeProto_AttributeType_STRING);
+    attr->set_s(def.engine());
+  }
+  if (def.control_input_size() > 0) {
+    auto attr = node->add_annotations();
+    attr->set_name("control_input");
+    attr->set_type(::torch::AttributeProto_AttributeType::
+                       AttributeProto_AttributeType_STRINGS);
+    for (int i = 0; i < def.control_input_size(); ++i) {
```
Comment (BIT-silence) [inline caffe2/utils/proto_convert.cc]: attr->mutable_strings()->CopyFrom(def.control_input());
or
for (const string& input : def.control_input()) {
  attr->add_strings(input);
}

File: caffe2/utils/proto_convert.cc [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    def->set_debug_info(node.doc_string());
+  }
+  for (int i = 0; i < node.annotations_size(); ++i) {
+    const auto& attr = node.annotations(i);
+    CAFFE_ENFORCE(attr.has_name());
+    if (attr.name() == "engine") {
+      CAFFE_ENFORCE(attr.has_s());
+      def->set_engine(attr.s());
+    } else if (attr.name() == "control_input") {
+      for (int j = 0; j < attr.strings_size(); ++j) {
```
Comment (BIT-silence) [inline caffe2/utils/proto_convert.cc]: ditto

File: caffe2/utils/proto_convert.h [hunk 6]
```diff
@@ -0,0 +1,22 @@
+#pragma once
```
Comment (BIT-silence) [inline caffe2/utils/proto_convert.h]: #ifndef CAFFE2_UTILS_PROTO_CONVERT_H_
#define CAFFE2_UTILS_PROTO_CONVERT_H_

#endif // CAFFE2_UTILS_PROTO_CONVERT_H_

**Review by xiaomengy — APPROVED** [review]
LGTM

File: caffe2/python/pybind_state.cc [hunk 7]
```diff
@@ -1186,11 +1187,17 @@ void addGlobalMethods(py::module& m) {
     return true;
   });
   m.def("nets", []() { return gWorkspace->Nets(); });
-  m.def("run_operator_once", [](const py::bytes& op_def) {
+  m.def("run_operator_once", [](const py::bytes& op_def, bool legacy_proto=true) {
```
Comment (dzhulgakov) [inline caffe2/python/pybind_state.cc]: hm, adding it to this API feels a bit out of place now. Is it just for test?

Comment (houseroad) [reply caffe2/python/pybind_state.cc]: Right now, it's just for test purpose.

**Pull request closed without merging**